File: ml/inference-gateway/internal/model/prediction.go
package model

// PredictionRequest mirrors contracts/schemas/prediction.json -> PredictionRequest.
type PredictionRequest struct {
	UserID                int     `json:"user_id" binding:"required,min=1"`
	EngagementScore       float64 `json:"engagement_score" binding:"required,min=0,max=1"`
	DaysSinceLogin        int     `json:"days_since_login" binding:"min=0"`
	SessionsLast30d       int     `json:"sessions_last_30d" binding:"min=0"`
	SupportTicketsLast90d int     `json:"support_tickets_last_90d" binding:"min=0"`
}

// PredictionResponse mirrors contracts/schemas/prediction.json -> PredictionResponse.
type PredictionResponse struct {
	UserID           int     `json:"user_id"`
	ChurnProbability float64 `json:"churn_probability"`
	RiskBand         string  `json:"risk_band,omitempty"`
	ModelVersion     string  `json:"model_version"`
}

File: ml/model-serving/app/schemas/prediction.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

RiskBand = Literal["low", "medium", "high"]


class PredictionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", protected_namespaces=())

    user_id: int = Field(..., ge=1)
    engagement_score: float = Field(..., ge=0.0, le=1.0)
    days_since_login: int = Field(..., ge=0)
    sessions_last_30d: int = Field(..., ge=0)
    support_tickets_last_90d: int = Field(..., ge=0)


class PredictionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", protected_namespaces=())

    user_id: int = Field(..., ge=1)
    churn_probability: float = Field(..., ge=0.0, le=1.0)
    risk_band: RiskBand
    model_version: str
